File: src/test/java/com/github/nsc/de/compiler/parser/IdentifierTests.java
package com.github.nsc.de.compiler.parser;

import com.github.nsc.de.compiler.parser.node.IdentifierNode;
import com.github.nsc.de.compiler.parser.node.Tree;
import com.github.nsc.de.compiler.parser.node.functions.FunctionCallNode;
import com.github.nsc.de.compiler.parser.node.variables.VariableUsageNode;
import org.junit.jupiter.api.Test;

import static com.github.nsc.de.compiler.TestUtil.assertType;
import static com.github.nsc.de.compiler.parser.ParserTestUtil.parse;
import static org.junit.jupiter.api.Assertions.*;

public class IdentifierTests {

    @Test
    public void testBasicIdentifier() {

        Tree tree = parse("<BasicIdentifierTest>", "test");
        assertEquals(1, tree.getChildren().length);
        assertType(VariableUsageNode.class, tree.getChildren()[0]);

        VariableUsageNode node = (VariableUsageNode) tree.getChildren()[0];
        assertEquals("test", node.getVariable().getName());
        assertNull(node.getVariable().getParent());

    }

    @Test
    public void testComplexIdentifier() {

        Tree tree = parse("<BasicIdentifierTest>", "test.test2");
        assertEquals(1, tree.getChildren().length);
        assertType(VariableUsageNode.class, tree.getChildren()[0]);

        VariableUsageNode node = (VariableUsageNode) tree.getChildren()[0];
        assertEquals("test2", node.getVariable().getName());
        assertNotNull(node.getVariable().getParent());
        assertType(VariableUsageNode.class, node.getVariable().getParent());

        node = (VariableUsageNode) node.getVariable().getParent();
        assertEquals("test", node.getVariable().getName());
        assertNull(node.getVariable().getParent());

    }

    @Test
    public void testComplexIdentifierWithFunctions() {

        Tree tree = parse("<BasicIdentifierTest>", "aaa.test().test2");
        assertEquals(1, tree.getChildren().length);
        assertType(VariableUsageNode.class, tree.getChildren()[0]);

        VariableUsageNode node = (VariableUsageNode) tree.getChildren()[0];
        assertEquals("test2", node.getVariable().getName());
        assertNotNull(node.getVariable().getParent());
        assertType(FunctionCallNode.class, node.getVariable().getParent());

        FunctionCallNode node2 = (FunctionCallNode) node.getVariable().getParent();
        assertNotNull(node.getVariable().getParent());
        assertType(IdentifierNode.class, node2.getFunction());
        assertEquals("test", ((IdentifierNode) node2.getFunction()).getName());
        assertType(VariableUsageNode.class, ((IdentifierNode) node2.getFunction()).getParent());

        node = (VariableUsageNode) ((IdentifierNode) node2.getFunction()).getParent();
        assertEquals("aaa", node.getVariable().getName());
        assertNull(node.getVariable().getParent());

    }
}

File: src/test/kotlin/com/github/nsc/de/shake/parser/IdentifierTests.kt
package com.github.nsc.de.shake.parser

import com.github.nsc.de.shake.assertType
import org.junit.jupiter.api.Test
import com.github.nsc.de.shake.parser.node.variables.VariableUsageNode
import com.github.nsc.de.shake.parser.node.functions.FunctionCallNode
import com.github.nsc.de.shake.parser.node.IdentifierNode
import org.junit.jupiter.api.Assertions

class IdentifierTests {
    @Test
    fun testBasicIdentifier() {
        val tree = ParserTestUtil.parse("<BasicIdentifierTest>", "test")
        Assertions.assertEquals(1, tree.children.size)
        assertType(VariableUsageNode::class.java, tree.children[0])
        val node = tree.children[0] as VariableUsageNode
        Assertions.assertEquals("test", node.variable.name)
        Assertions.assertNull(node.variable.parent)
    }

    @Test
    fun testComplexIdentifier() {
        val tree = ParserTestUtil.parse("<BasicIdentifierTest>", "test.test2")
        Assertions.assertEquals(1, tree.children.size)
        assertType(VariableUsageNode::class.java, tree.children[0])
        var node = tree.children[0] as VariableUsageNode
        Assertions.assertEquals("test2", node.variable.name)
        Assertions.assertNotNull(node.variable.parent)
        assertType(VariableUsageNode::class.java, node.variable.parent)
        node = node.variable.parent as VariableUsageNode
        Assertions.assertEquals("test", node.variable.name)
        Assertions.assertNull(node.variable.parent)
    }

    @Test
    fun testComplexIdentifierWithFunctions() {
        val tree = ParserTestUtil.parse("<BasicIdentifierTest>", "aaa.test().test2")
        Assertions.assertEquals(1, tree.children.size)
        assertType(VariableUsageNode::class.java, tree.children[0])
        var node = tree.children[0] as VariableUsageNode
        Assertions.assertEquals("test2", node.variable.name)
        Assertions.assertNotNull(node.variable.parent)
        assertType(FunctionCallNode::class.java, node.variable.parent)
        val node2 = node.variable.parent as FunctionCallNode
        Assertions.assertNotNull(node.variable.parent)
        assertType(IdentifierNode::class.java, node2.function)
        Assertions.assertEquals("test", (node2.function as IdentifierNode).name)
        assertType(VariableUsageNode::class.java, (node2.function as IdentifierNode).parent)
        node = (node2.function as IdentifierNode).parent as VariableUsageNode
        Assertions.assertEquals("aaa", node.variable.name)
        Assertions.assertNull(node.variable.parent)
    }
}